components/raftstore-v2/src/operation/ready/snapshot.rs

```diff
@@ -0,0 +1,275 @@
+    /// asynchronously.
+    pub fn handle_snapshot(&mut self, snap_task: GenSnapTask) {
+        // Flush before do snapshot.
+        self.flush();
+
+        // Send generate snapshot task to region worker.
+        let (last_applied_index, last_applied_term) = self.apply_progress();
+        snap_task.index.store(last_applied_index, Ordering::SeqCst);
+        let gen_tablet_sanp_task = ReadTask::GenTabletSnapshot {
+            region_id: snap_task.region_id,
+            tablet: self.tablet().clone(),
+            region_state: self.region_state().clone(),
+            last_applied_term,
+            last_applied_index,
+            for_balance: snap_task.for_balance,
+            canceled: snap_task.canceled.clone(),
+        };
+        if let Err(e) = self.read_scheduler().schedule(gen_tablet_sanp_task) {
+            error!(
+                self.logger,
+                "schedule snapshot failed";
+                "error" => ?e,
+            );
+            snap_task.canceled.store(true, Ordering::SeqCst);
+        }
+    }
+}
+
+impl<EK: KvEngine, ER: RaftEngine> Storage<EK, ER> {
+    /// Gets a snapshot. Returns `SnapshotTemporarilyUnavailable` if there is no
+    /// unavailable snapshot.
+    pub fn snapshot(&self, request_index: u64, to: u64) -> raft::Result<Snapshot> {
+        let mut snap_state = self.snap_state_mut();
+        match *snap_state {
+            SnapState::Generating { ref canceled, .. } => {
+                if canceled.load(Ordering::SeqCst) {
+                    self.cancel_generating_snap(None);
+                } else {
+                    return Err(raft::Error::Store(
+                        raft::StorageError::SnapshotTemporarilyUnavailable,
+                    ));
+                }
+            }
+            SnapState::Generated(ref s) => {
+                let SnapState::Generated(snap) = mem::replace(&mut *snap_state, SnapState::Relax) else { unreachable!() };
+                if self.validate_snap(&snap, request_index) {
+                    return Ok(*snap);
+                }
+            }
+            _ => {}
+        }
+
+        if SnapState::Relax != *snap_state {
+            panic!(
+                "{:?} unexpected state: {:?}",
+                self.logger().list(),
+                *snap_state
+            );
+        }
+
+        info!(
+            self.logger(),
+            "requesting snapshot";
+            "tablet_index" => self.get_tablet_index(),
+            "request_index" => request_index,
+            "request_peer" => to,
+        );
+        let canceled = Arc::new(AtomicBool::new(false));
+        let index = Arc::new(AtomicU64::new(0));
+        *snap_state = SnapState::Generating {
+            canceled: canceled.clone(),
+            index: index.clone(),
+        };
+
+        let task = GenSnapTask::new(self.region().get_id(), index, canceled);
+        let mut gen_snap_task = self.gen_snap_task_mut();
+        assert!(gen_snap_task.is_none());
```

Comment:
If:

0. raft calls snapshot()
1. Somehow peer cancel snapshot
2. raft calls snapshot() again
3. handle ready

This line may panic.

Reply:
added logic `takethe task` in cancel.